lib/model/unreads.dart

```diff
@@ -0,0 +1,411 @@
+        // as flags. Instead, we just get the [UpdateMessageEvent], and that
+        // contains the new set of flags, which we'll use to update [mentions].
+        // (See our handling of [UpdateMessageEvent].)
+        //
+        // Handle the event anyway, using the meaning on the tin.
+        // It might be used in a valid case we haven't thought of yet.
+
+        // TODO skip notifyListeners if unchanged?
+        switch (event) {
+          case UpdateMessageFlagsAddEvent():
+            mentions.addAll(
+              event.messages.where(
+                (messageId) => _slowIsPresentInStreams(messageId) || _slowIsPresentInDms(messageId),
+              ),
+            );
+
+          case UpdateMessageFlagsRemoveEvent():
+            mentions.removeAll(event.messages);
+        }
+
+      case MessageFlag.read:
+        switch (event) {
+          case UpdateMessageFlagsAddEvent():
+            if (event.all) {
+              streams.clear();
+              dms.clear();
+              mentions.clear();
+              oldUnreadsMissing = false;
+            } else {
+              final messageIdsSet = Set.of(event.messages);
+              mentions.removeAll(messageIdsSet);
+              _slowRemoveAllInStreams(messageIdsSet);
+              _slowRemoveAllInDms(messageIdsSet);
+            }
+          case UpdateMessageFlagsRemoveEvent():
+            final newlyUnreadInStreams = <int, Map<String, QueueList<int>>>{};
+            final newlyUnreadInDms = <DmNarrow, QueueList<int>>{};
+            for (final messageId in event.messages) {
+              final detail = event.messageDetails![messageId];
+              if (detail == null) { // TODO(log) if on Zulip 6.0+
+                // Happens as a bug in some cases before fixed in Zulip 6.0:
+                //   https://chat.zulip.org/#narrow/stream/378-api-design/topic/unreads.20in.20unsubscribed.20streams/near/1458467
+                // TODO(server-6) remove Zulip 6.0 comment
+                continue;
+              }
+              if (detail.mentioned == true) {
+                mentions.add(messageId);
+              }
+              switch (detail.type) {
+                case MessageType.stream:
+                  final perStream = (newlyUnreadInStreams[detail.streamId!] ??= {});
+                  final perTopic = (perStream[detail.topic!] ??= QueueList());
+                  perTopic.add(messageId);
+                case MessageType.private:
+                  final narrow = DmNarrow.ofUpdateMessageFlagsMessageDetail(selfUserId: selfUserId,
+                    detail);
+                  (newlyUnreadInDms[narrow] ??= QueueList())
+                    .add(messageId);
+              }
+            }
+            newlyUnreadInStreams.forEach((incomingStreamId, incomingPerTopic) {
+              incomingPerTopic.forEach((incomingTopic, incomingMessageIds) {
+                _addAllInStreamTopic(incomingMessageIds..sort(), incomingStreamId, incomingTopic);
```

Comment:
Hmm, this uses "per topic" with a different meaning from in the code above:
```dart
                  final perStream = (newlyUnreadInStreams[detail.streamId!] ??= {});
                  final perTopic = (perStream[detail.topic!] ??= QueueList());
                  perTopic.add(messageId);
```

The `incomingPerTopic` here corresponds to the `perStream` there.

Perhaps that means that "per topic" is too ambiguous — so possibly best to avoid it on both sides. But in any case should avoid using it with two contrasting meanings. :slightly_smiling_face: